### PR TITLE
Starts using AutoValue for classes in stats-core.

### PR DESCRIPTION
**Pull request by sebright**
This PR applies AutoValue to `MeasurementDescriptor` and `MeasurementUnit` and adds some tests.

### CODE REVIEW

**Comment by dinooliva** [comment]
Before approving this change, I need to understand the implications of going to autovalue. A few questions that come to mind:
- How well supported is it by non-bazel build systems?
- Is there any performance overhead compared to the hand-written class?
- If we are going this route for MeasurementDescriptor, shouldn't we consistently do this for the other View-related data structures?

**Comment by sebright** [comment]
I still need to debug the warning in the Bazel build.

**Comment by sebright** [comment]
I added a commit to fix the build:

#### Suppress annotation processor warning in Bazel CI build.

This commit suppresses a javac "processing" warning that was enabled with
`--javacopt=-Xlint:all`.  The warning was `"warning: No processor claimed any of
these annotations: com.google.auto.value.AutoValue,javax.annotation.Nullable"`
According to the response in
https://groups.google.com/forum/#!topic/bazel-discuss/_R3A9TJSoPM, the best
solution is to ignore the warning.

File: pom.xml [hunk 1]
```diff
@@ -170,6 +170,12 @@
       <version>3.0.0</version>
     </dependency>
     <dependency>
+      <groupId>com.google.auto.value</groupId>
+      <artifactId>auto-value</artifactId>
+      <version>1.3</version>
```
Comment (adriancole) [inline pom.xml]: 1.4 is latest (now :) )

**Comment by sebright** [comment]
@adriancole Thanks! I upgraded to 1.4.

**Comment by sebright** [comment]
@bogdandrutu @dinooliva Do you mind if I merge this?

File: core/src/main/java/com/google/instrumentation/stats/MeasurementDescriptor.java [hunk 2]
```diff
@@ -22,47 +24,34 @@
  *
  * <p>Note: MeasurementDescriptor names are {@link String}s with enforced restrictions.
  */
-public final class MeasurementDescriptor {
+@AutoValue
+public abstract class MeasurementDescriptor {
   public static final int MAX_LENGTH = StringUtil.MAX_LENGTH;
 
+  MeasurementDescriptor() {}
```
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/stats/MeasurementDescriptor.java]: Please add a comment:
// Disallow extending this class outside this package ?

Comment (sebright) [inline core/src/main/java/com/google/instrumentation/stats/MeasurementDescriptor.java]: I thought that that type of comment would become redundant if we did the same thing for all value classes.  Do you think it is still worth documenting within each class?

Comment (sebright) [inline core/src/main/java/com/google/instrumentation/stats/MeasurementDescriptor.java]: We discussed this in person and decided against the documentation.

**Comment by sebright** [comment]
Thanks for the reviews!

**Pull request merged**